### PR TITLE
Add resource interpreter document

### LINKED ISSUE #1250: [docs] Customizing resource interpreter
**What would you like to be added**:
Add documents about [ResourceInterpreter](https://github.com/karmada-io/karmada/tree/master/docs/proposals/resource-interpreter-webhook) at `pkg/docs/userguide/customizing-resource-interpreter.md`.

**Why is this needed**:

### PR DESCRIPTION
**Pull request by RainbowMango**
**What type of PR is this?**
/kind documentation

**What this PR does / why we need it**:
Added `CustomResourceInterpreter` document.

**Which issue(s) this PR fixes**:
Fixes #1250

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by RainbowMango** [comment]
@Tingtal Could you please help take a look?

File: docs/userguide/customizing-resource-interpreter.md [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+## Resource Interpreter Framework
+
+In the progress of propagating a resource from `karmada-apiserver` to member clusters, Karmada needs to know the 
+resource definition. Take `Propagating Deployment` as an example, at the phase of building `ResourceBinding`, the 
+`karmada-controller-manager` will parse the `replicas` from the deployment object.
+
+For Kubernetes native resources, Karmada knows how to parse them, but for custom resources defined by `CRD`(or extended
+by something like `aggregated-apiserver`), as lack of the knowledge of the resource structure, they can only be treated 
+as normal resources, therefore, the advanced scheduling algorithms cannot be used for them.
```
Comment (Tingtal) [inline docs/userguide/customizing-resource-interpreter.md:31]:
```suggestion
as normal resources. Therefore, the advanced scheduling algorithms cannot be used for them.
```

File: docs/userguide/customizing-resource-interpreter.md [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+The [Resource Interpreter Framework][1] is designed for interpreting resource structure. It consists of `built-in` and 
+`customized` interpreters:
+- built-in interpreter: built-in interpreters for common Kubernetes native or well-known extended resources.
+- customized interpreter: Used to interpreter custom resources or override the built-in interpreters.
```
Comment (Tingtal) [inline docs/userguide/customizing-resource-interpreter.md:36]:
```suggestion
- customized interpreter: interprets custom resources or overrides the built-in interpreters.
```

File: docs/userguide/customizing-resource-interpreter.md [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+> Note: The main difference between `built-in` and `customized` interpreter is the `built-in` interpreted is
```
Comment (Tingtal) [inline docs/userguide/customizing-resource-interpreter.md:38]:
```suggestion
> Note: The major difference between `built-in` and `customized` interpreters is that the `built-in` interpreter is 
```

File: docs/userguide/customizing-resource-interpreter.md [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+by something like `aggregated-apiserver`), as lack of the knowledge of the resource structure, they can only be treated 
+as normal resources, therefore, the advanced scheduling algorithms cannot be used for them.
+
+The [Resource Interpreter Framework][1] is designed for interpreting resource structure. It consists of `built-in` and 
+`customized` interpreters:
+- built-in interpreter: built-in interpreters for common Kubernetes native or well-known extended resources.
+- customized interpreter: Used to interpreter custom resources or override the built-in interpreters.
+
+> Note: The main difference between `built-in` and `customized` interpreter is the `built-in` interpreted is 
+> implemented and maintained by Karmada community, it will be built into Karmada components, such as
```
Comment (Tingtal) [inline docs/userguide/customizing-resource-interpreter.md:39]:
```suggestion
> implemented and maintained by Karmada community and will be built into Karmada components, such as 
```

File: docs/userguide/customizing-resource-interpreter.md [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+as normal resources, therefore, the advanced scheduling algorithms cannot be used for them.
+
+The [Resource Interpreter Framework][1] is designed for interpreting resource structure. It consists of `built-in` and 
+`customized` interpreters:
+- built-in interpreter: built-in interpreters for common Kubernetes native or well-known extended resources.
+- customized interpreter: Used to interpreter custom resources or override the built-in interpreters.
+
+> Note: The main difference between `built-in` and `customized` interpreter is the `built-in` interpreted is 
+> implemented and maintained by Karmada community, it will be built into Karmada components, such as 
+> `karmada-controller-manager`. On the contrary, the `customized` interpreter is implemented and maintained by users,
```
Comment (Tingtal) [inline docs/userguide/customizing-resource-interpreter.md:40]:
```suggestion
> `karmada-controller-manager`. On the contrary, the `customized` interpreter is implemented and maintained by users.
```

File: docs/userguide/customizing-resource-interpreter.md [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+The [Resource Interpreter Framework][1] is designed for interpreting resource structure. It consists of `built-in` and 
+`customized` interpreters:
+- built-in interpreter: built-in interpreters for common Kubernetes native or well-known extended resources.
+- customized interpreter: Used to interpreter custom resources or override the built-in interpreters.
+
+> Note: The main difference between `built-in` and `customized` interpreter is the `built-in` interpreted is 
+> implemented and maintained by Karmada community, it will be built into Karmada components, such as 
+> `karmada-controller-manager`. On the contrary, the `customized` interpreter is implemented and maintained by users, 
+> it should be registered to Karmada as an `Interpreter Webhook`(See below for more details).
```
Comment (Tingtal) [inline docs/userguide/customizing-resource-interpreter.md:41]:
```suggestion
> It should be registered to Karmada as an `Interpreter Webhook` (see below for more details).
```

File: docs/userguide/customizing-resource-interpreter.md [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+### Interpreter Operations
+
+When interpreting resource, we usually to extract multiple pieces of information from it, the `Interpreter Operations`
```
Comment (Tingtal) [inline docs/userguide/customizing-resource-interpreter.md:45]:
```suggestion
When interpreting resources, we often get multiple pieces of information extracted. The `Interpreter Operations`
```

File: docs/userguide/customizing-resource-interpreter.md [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+defines the interpreter request type, and the `Resource Interpreter Framework` provides services for each operation 
+type. 
+
+For all operations designed by `Resource Interpreter Framework`, please refer to [Interpreter Operations][2].
+
+> Note: Not all the designed operations are supported(See below for supported operations).
```
Comment (Tingtal) [inline docs/userguide/customizing-resource-interpreter.md:51]:
```suggestion
> Note: Not all the designed operations are supported (see below for supported operations).
```

File: docs/userguide/customizing-resource-interpreter.md [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+
+> Note: At most one interpreter will be consulted to when interpreting a resource with specific `interpreter operation`
+> and the `customized` interpreter has higher priority than `built-in` interpreter if they both interpreting the same
```
Comment (Tingtal) [inline docs/userguide/customizing-resource-interpreter.md:54]:
```suggestion
> and the `customized` interpreter has higher priority than `built-in` interpreter if they are both interpreting the same 
```

File: docs/userguide/customizing-resource-interpreter.md [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+type. 
+
+For all operations designed by `Resource Interpreter Framework`, please refer to [Interpreter Operations][2].
+
+> Note: Not all the designed operations are supported(See below for supported operations).
+
+> Note: At most one interpreter will be consulted to when interpreting a resource with specific `interpreter operation`
+> and the `customized` interpreter has higher priority than `built-in` interpreter if they both interpreting the same 
+> resource. 
+> For example, the `built-in` interpreter serves `InterpretReplica` for `Deployment` with version `apps/v1`, if there
```
Comment (Tingtal) [inline docs/userguide/customizing-resource-interpreter.md:56]:
```suggestion
> For example, the `built-in` interpreter serves `InterpretReplica` for `Deployment` with version `apps/v1`. If there 
```

File: docs/userguide/customizing-resource-interpreter.md [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+
+For all operations designed by `Resource Interpreter Framework`, please refer to [Interpreter Operations][2].
+
+> Note: Not all the designed operations are supported(See below for supported operations).
+
+> Note: At most one interpreter will be consulted to when interpreting a resource with specific `interpreter operation`
+> and the `customized` interpreter has higher priority than `built-in` interpreter if they both interpreting the same 
+> resource. 
+> For example, the `built-in` interpreter serves `InterpretReplica` for `Deployment` with version `apps/v1`, if there 
+> is a customized interpreter registered to Karmada for interpreting the same resource, the `customized` interpreter win and the
```
Comment (Tingtal) [inline docs/userguide/customizing-resource-interpreter.md:57]:
```suggestion
> is a customized interpreter registered to Karmada for interpreting the same resource, the `customized` interpreter wins and the 
```

File: docs/userguide/customizing-resource-interpreter.md [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+> Note: At most one interpreter will be consulted to when interpreting a resource with specific `interpreter operation`
+> and the `customized` interpreter has higher priority than `built-in` interpreter if they both interpreting the same 
+> resource. 
+> For example, the `built-in` interpreter serves `InterpretReplica` for `Deployment` with version `apps/v1`, if there 
+> is a customized interpreter registered to Karmada for interpreting the same resource, the `customized` interpreter win and the 
+> `built-in` interpreter will be ignored.
+
+## Built-in Interpreter
+
+For the common Kubernetes native or well-known extended resources, the interpreter operations are built-in, that means
```
Comment (Tingtal) [inline docs/userguide/customizing-resource-interpreter.md:62]:
```suggestion
For the common Kubernetes native or well-known extended resources, the interpreter operations are built-in, which means
```

File: docs/userguide/customizing-resource-interpreter.md [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+the users usually don't need to implement the customized interpreter. If you want more resources should be built-in,
```
Comment (Tingtal) [inline docs/userguide/customizing-resource-interpreter.md:63]:
```suggestion
the users usually don't need to implement customized interpreters. If you want more resources to be built-in,
```

File: docs/userguide/customizing-resource-interpreter.md [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+
+### AggregateStatus
+
+Supported resources:
+- Deployment(apps/v1)
+- Service(v1)
+- Ingress(extensions/v1beta1)
+- Job(batch/v1)
+
+## Customize Interpreter
```
Comment (Tingtal) [inline docs/userguide/customizing-resource-interpreter.md:97]:
```suggestion
## Customized Interpreter
```

File: docs/userguide/customizing-resource-interpreter.md [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+
+The customized interpreter is implemented and maintained by users, it developed as extensions and
```
Comment (Tingtal) [inline docs/userguide/customizing-resource-interpreter.md:99]:
```suggestion
A customized interpreter is implemented and maintained by users, developed as an extension and 
```

File: docs/userguide/customizing-resource-interpreter.md [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+Supported resources:
+- Deployment(apps/v1)
+- Service(v1)
+- Ingress(extensions/v1beta1)
+- Job(batch/v1)
+
+## Customize Interpreter
+
+The customized interpreter is implemented and maintained by users, it developed as extensions and 
+run as webhooks at runtime.
```
Comment (Tingtal) [inline docs/userguide/customizing-resource-interpreter.md:100]:
```suggestion
running as a webhook at runtime.
```

File: docs/userguide/customizing-resource-interpreter.md [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+
+### What are interpreter webhooks?
+
+Interpreter webhooks are HTTP callbacks that receive interpret requests and do something with them.
+
+### Write an interpreter webhook server
+
+Please refer to the implementation of the [Example of Customize Interpreter][4] that is validated 
+in Karmada E2E test. The webhook handles the `ResourceInterpreterRequest` request sent by the 
+Karmada components(such as `karmada-controller-manager`), and sends back its decision as an
```
Comment (Tingtal) [inline docs/userguide/customizing-resource-interpreter.md:110]:
```suggestion
Karmada components (such as `karmada-controller-manager`), and sends back its decision as an 
```

File: docs/userguide/customizing-resource-interpreter.md [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+
+The [Example of Customize Interpreter][4] is deployed in the host cluster for E2E and exposed by 
+a service as the front-end of the webhook server.
+
+You may also deploy your webhooks outside the cluster. You will need to update your webhook 
+configurations accordingly.
+
+### Configure webhook on the fly
+
+You can configure what resources and supported operations are subject to what interpreter webhook
```
Comment (Tingtal) [inline docs/userguide/customizing-resource-interpreter.md:123]:
```suggestion
You can configure which resources and supported operations are subject to which interpreter webhook 
```

**Review by Tingtal — CHANGES_REQUESTED** [review]

**Pull request merged**